### ISSUE TITLE
Simulations with differen corner-library (`.lib ../.../cornerMOSlv_psp.lib mos_tt and ...ss ...ff ...fs ...sf`), what would You do?

### ISSUE THREAD

**Issue by StefanSchippers** [StefanSchippers/xschem#144]
@olisnr

> if You want to make some simulations with differen corner-library (`.lib ../.../cornerMOSlv_psp.lib mos_tt and ...ss ...ff ...fs ...sf`), what would You do? 
> i can use a `launcher.sym` to set a lib-variable and then `xschem netlist; xschem simulate` but i didnt figured out, how to wait of the end of the simulation before update the lib-variable, and simulate again.
> or i could use TCL to write different spice-netlists and run them in parallel (also without knowing when it ended)...
> 
> now it looks like:
> 
> ![corner](https://github.com/StefanSchippers/xschem/assets/120017377/cc322dcf-d261-4dfa-a8d4-f4bb3a009ecd)
> 
> and the possibility to load different .raw-files is REALLY great!
> 
> _Originally posted by @olisnr in https://github.com/StefanSchippers/xschem/issues/141#issuecomment-1816409510_

**Comment by StefanSchippers** [comment]
Opened a new issue-thread since the other grew too long.

Waiting for a simulation (or some )to finish is not difficult. 
If you want to run more in parallel (this is ideal for corner sims since these run can go independently) you need to ensure outut files go in different places. You can define some tcl variables in the code blocks (and use tcleval(...) ).

Then to run some sims and wait do the following:

```
#set tcl variables for 1st run
xschem netlist
set id1 [simulate]

#set tcl variables for 2nd run
xschem netlist
set id2 [simulate]

#set tcl variables for 3rd run
xschem netlist
set id2 [simulate]

# wait for completion of all runs.If they are different corners they should finish in similar times

vwait execute(cmd,$id1)
vwait execute(cmd,$id2)
vwait execute(cmd,$id3)
```
execute(cmd,id) is the command line of the running process. This variable (and others related to this simulation) get deleted at the end (except the new `execute(exitcode,id)` ) . so we can vwait for them (vwait blocks until something happens with the variable).

PS: `simulate` is the recommended way to launch simulations instead of the legacy `xschem simulate`.
the simulate procedure returns an integer id to use with the execute variables:

```
# $execute(id) is an integer identifying the last running pipeline
#       set id $execute(id) to get the current subprocess id in variable 'id'
# $execute(status,$id) contains the status argument as given in proc execute
# $execute(pipe,$id) contains the channel descriptor to read or write as specified in status
# $execute(data,$id) contains the stdout of the pipeline (output data)
# $execute(cmd,$id) contains the pipeline command
# $execute(win_path,$id) contains the xctx->current_win_path that started the command
# $execute(exitcode,$id) contains the pipeline command (this is a new addition in xschem)
```

**Comment by StefanSchippers** [comment]
Following an old request where you were asking to use tcl variables in node names / datasets / whatever in the graph dialog, the new approach is the following. Instead of wrapping some parts with tcleval() just wrap the entire list of nodes inside a tcleval(....) as shwn below. The second node is completely specified by fariables (alias name, signal, dataset, raw file):

![1](https://github.com/StefanSchippers/xschem/assets/69359491/1ecb8a37-bc22-42ea-b9c7-b43a9f4be721)

**Comment by olisnr** [comment]
many thanks! i was trying with the $execute(...) but it wasnt working. 
the parallel code looks super, i will test it.

and im very happy with plotting different simulations together.

do You have an idea, how one could annotate an already open sub-sheet? or go back to the simulation controlling top-sheet without closing the sub-sheet?

**Comment by StefanSchippers** [comment]
> many thanks! i was trying with the $execute(...) but it wasnt working. the parallel code looks super, i will test it.
> 
> and im very happy with plotting different simulations together.
> 
> do You have an idea, how one could annotate an already open sub-sheet? or go back to the simulation controlling top-sheet without closing the sub-sheet?

May be in the parallel code above you need to add a delay (after 1000) . If ngspice has not yet read the netlist and you generate the netlist for the second run you get a mess.
Another solution is to use different netlist names (xschem set netlist_name xxxx.spice)  for the runs, so you avoid such race conditions.

**Comment by StefanSchippers** [comment]
> do You have an idea, how one could annotate an already open sub-sheet? or go back to the simulation controlling top-sheet without closing the sub-sheet?

Some additional work needs to be done, but you can annotate now a subsheet without leaving the subsheet. See video clip.


https://github.com/StefanSchippers/xschem/assets/69359491/847b1068-153c-4877-9b6f-1594b8da9dcb

**Comment by olisnr** [comment]
thats great. is there a way to get the name of the top-sheet from the child? then i could make an universal annotate `launcher.sym` 

in the [doc](http://repo.hu/projects/xschem/xschem_man/xschem_man.html) i didnt found something.

**Comment by olisnr** [comment]
i tried the child annotation, but something didnt work correct:


https://github.com/StefanSchippers/xschem/assets/120017377/78174102-0804-44e6-a3c7-6a97aefaae27

**Comment by StefanSchippers** [comment]
Yes you can get the name (full path) of parent levels with
`xschem get schname n`
where n is the hierarchy level (0= top)

**Comment by StefanSchippers** [comment]
> i tried the child annotation, but something didnt work correct:

It did not work since you opened the sub sheet with Alt-E. This just opens a new top level. 
Do the following steps.
- open again the top schematic .
- descend with 'e' into the subsheet.
- open a new empty tab
- open the top level. You can use this top level to run /  change simulations.
- the tab with the subsheet should be able to annotate the OP data with your launcher.

**Comment by olisnr** [comment]
the launcher works after `e` using:
`tclcommand="xschem annotate_op ${netlist_dir}/[file rootname [file tail [xschem get schname 0]]].raw 0"`

thanks!

**Comment by olisnr** [comment]
maybe the doc for 
```
raw_clear
   Delete loaded simulation raw file
```
should be `unload` instead of `delete`

**Comment by olisnr** [comment]
some other thing i would like in the documentation:

_in a tcleval() we do a tcl command and variable substitution. This means inside the tcleval() there is a string. The result is the string with tcl variable substitution (any $var or ${var} (better) instances will be replaced with the var content) and command substitution (any [command] instances will be replaced with the return value of the tcl command). Of course tcl commands can do various side effects, like setting variables, printing stuff, even creating new procedures. The tcleval() is processed by removing the tcleval( and the ending ), then passing the string to the tcl subst command.

A tclcommand attribute in the launcher is just taken as is (no tcleval() needed) and passed to the eval tcl function. So for example in a launcher you don't need to wrap the command into [...].

The tcleval() construct is done also for efficiency reasons. If an attribute does not begin wth tcleval() no call to the tcl interpreter is done. A C strstr() check on the tcleval( string is super fast.

The difference between eval and subst is that any string is valid for subst (if no commands or variables are found it is returned as is, and this is what we want for string attributes. tclcommand attributes are directly passed to the eval function and must be valid tcl programs just as you are typing them at the tcl prompt. All the executions mentioned are done in the global scope._

**Comment by olisnr** [comment]
im trying the multithreading code, but i have problems with it. 
now it looks like:

```
tclcommand="
	set corners [list tt ss ff sf fs]

	foreach corner $corners { ### debug
		puts $corner
		puts [string cat {$id_} ${corner}]
	}

	foreach corner $corners {
		set hv_corner $corner; 
		xschem netlist; 
		delay 111;
		set [string cat {$id_} ${corner}] [simulate]; 
	}

	foreach corner $corners {
		puts [string cat {$execute(cmd, } [string cat {$id_} ${corner}] \)]; 
		vwait [string cat {$execute(cmd, } [string cat {$id_} ${corner}] \)]; 
	}


about ### debug

	xschem raw_clear;
	xschem raw_read [rawfilestart].raw; 
	xschem annotate_op;
"
```

it says:
```
tt
$id_tt
ss
$id_ss
ff
$id_ff
sf
$id_sf
fs
$id_fs
Simulation started: execution ID: 178
Simulation started: execution ID: 179
Simulation started: execution ID: 180
Simulation started: execution ID: 181
Simulation started: execution ID: 182
$execute(cmd, $id_tt)

```

it simulates something, but hangs vor ever in
`vwait [string cat {$execute(cmd, } [string cat {$id_} ${corner}] \)]; ` ( `vwait $execute(cmd, $id_tt)` )...

i didnt found vwait in the doc, and im not really sure how it should work

**Comment by StefanSchippers** [comment]
vwait is a tcl command so get the info with man winfo.
also remove the space after cmd, in these lines:
```
		puts [string cat {$execute(cmd, } [string cat {$id_} ${corner}] \)]; 
		vwait [string cat {$execute(cmd, } [string cat {$id_} ${corner}] \)]; 
```
as `execute(cmd, n)` and `execute(cmd,n)` (where n is an integer) are 2 different variables. In the first case you are waiting fore a variable that does not exists and so forever.

**Comment by StefanSchippers** [comment]
I think also the:
```
vwait [string cat {$execute(cmd, } [string cat {$id_} ${corner}] \)]
```
should be:
```
eval vwait [string cat {$execute(cmd,} [string cat {$id_} ${corner}] {)}]
```
since the variable you build with string cat needs to be substituted

**Comment by olisnr** [comment]
yes i changed it, now it produces `execute(cmd,$id_tt)` but still stalls. there is also something other wrong.

**Comment by StefanSchippers** [comment]
I was wrong, it seems you need to remove the `$` before `execute` ad the space after `cmd,`

**Comment by StefanSchippers** [comment]
also so a test with one corner only

**Comment by StefanSchippers** [comment]
Try this command on a circuit that takes some seconds to simulate. Does this work? (to me it does):
``` 
set id [simulate]; vwait [string cat {execute(cmd,} $id {)}]
```

**Comment by olisnr** [comment]
Your line is working, and my code is also working for only one thread iF i use the id as number and not as variable:
```
		vwait [string cat {execute(cmd,} [string cat {$id_} ${corner}] \)]; # blocks vor ever
		vwait [string cat {execute(cmd,} [set [string cat {id_} ${corner}]] \)]; # works...
```
if i start all simulations, then mostly it stalls after the 2nd vwait, but sometimes it event riches the 4th.
may be its a question of the ending times of the different threads

**Comment by StefanSchippers** [comment]
Yes I think something must be implemented here... if a vwait on one variable loses a write on another variable the second vwait stalls forever. Will think for a simpler solution. May be you can test a polling loop:
```
while {1} {
  after 1000 ;# check every second
  if { ![info exists var1] && ![info exists var2] } {
    # all variables removed, jobs finished
    break
}
```

something like that. var1 and var2 are the execute(cmd,n) variables.

**Comment by olisnr** [comment]
it wanted it in a loop:
```
	while {true} {
		set info_exists false;
		foreach corner $corners {
			set info_exists expr {$info_exists || [info exists [set [string cat {id_} ${corner}]]]}
		}
		if {!info_exists} {
			break;
		}
		delay 333
	}
```
but the `set info_exists expr {$info_exists || [info exists [set [string cat {id_} ${corner}]]]}` dont work. 
even `set info_exists expr {$info_exists}` isnt possible. why? `expr {$info_exists}` returns a false, but i cant load the variable whit it...

**Comment by StefanSchippers** [comment]
what is the output of 
```
string cat {id_} ${corner}
```
?

**Comment by StefanSchippers** [comment]
I will do something like this:
```
# run simulations
foreach {...corners...} {
  # set corner variables
  xschem netlist
  set id [simulate]
  lappend id_list $id
  delay 1000 ;# let simulator the time to read netlist
}

# poll loop
while {1} {
  delay 1000
  set busy 0
  foreach id $id_list {
    set var "execute(cmd,$id}"
    if {[info exists $var]} { set busy 1}
  }
  if {!$busy} { break }
}
puts Done
```

**Comment by StefanSchippers** [comment]
I have implemented the "true descend sub schematic in new tab/window" (the Alt-E key). See video.

https://github.com/StefanSchippers/xschem/assets/69359491/b6254710-e256-4ec1-ace8-8290f1c37b35

**Comment by olisnr** [comment]
super! i love this!!!

when make a new simulation, then i guess i have to reload the sub-sheets individual, or are the updated data used for all plots automatic?

**Comment by olisnr** [comment]
Xschem enjoys the address range of 32bit :)

![mem](https://github.com/StefanSchippers/xschem/assets/120017377/b2f65a15-a6c1-410a-a997-f3c75b0bdfbd)

**Comment by StefanSchippers** [comment]
> super! i love this!!!
> 
> when make a new simulation, then i guess i have to reload the sub-sheets individual, or are the updated data used for all plots automatic?

Currently you have to reload the data in the subsheets.
It could be possible in principle to share a single raw file between the top level and the subsheets, but this is a complex thing to handle. In general additional tabs or windows (in the same xschem process) are designed to be independent. What happens in one schematic on tab 'x' does not cause any side effect on another schematic in tab 'y'.

**Comment by StefanSchippers** [comment]
> Xschem enjoys the address range of 32bit :)
> 
> ![mem](https://user-images.githubusercontent.com/120017377/284410175-b2f65a15-a6c1-410a-a997-f3c75b0bdfbd.png)

how much raw files did you load?

**Comment by olisnr** [comment]
15

**Comment by StefanSchippers** [comment]
> 15

yes but total file size?

ngspice stores data as doubles (8 bytes / node/point) xschem converts to floats (4 bytes/ node/point) so the total size in memory should be 50%of the total file size. 7 significant digits of floats is more than enough in 99.995% of all simulations. There is though a compile option in xschem to keep data as doubles in memory.

**Comment by olisnr** [comment]
the data are 153MB on disk, and a new openened  Xschem needs 143MB RAM for it. i was trying to get the TCL-scripts working, and at some time there i saw the memory consumption, i had may errors in my scripts, and may be there where data loaded or generated and never unloaded of
may be there is a memory hole or two :)

**Comment by olisnr** [comment]
for the float / double question: my experience is, if You really want to do something whit the data: integrating, IQ demodulation, filtering or the like, float produces very fast big errors. i would prefer double, if You want to do something more then plotting in future.

**Comment by StefanSchippers** [comment]
> for the float / double question: my experience is, if You really want to do something whit the data: integrating, IQ demodulation, filtering or the like, float produces very fast big errors. i would prefer double, if You want to do something more then plotting in future.

you can change this line:
```
#define SPICE_DATA float
```
to double in xschem.h.
Consider however that simulation data has 3-4 significant digits.

**Comment by StefanSchippers** [comment]
> also the plots get strange peaks (that seems to be not real):

the AC plots are broken because all 5 simulation processes write into the same [rawfilestart]_ac.raw file. Add corner info like you did for the other files. also op data is written by 5 processes into the same [rawfilestart].raw file.
also there is a write done into [rawfilestart]_dc.raw
you must differentiate these output files per single process.

**Comment by StefanSchippers** [comment]
I have changed the schematic to avoid concurrent ngspice writes into the same raw files.
With all data loaded the xschem process uses 99MB of resident memory, and plots seem to be correct (no spikes)
![1](https://github.com/StefanSchippers/xschem/assets/69359491/5605cf26-92d6-46f5-9730-0e9f4fe1caf5)
[stef.zip](https://github.com/StefanSchippers/xschem/files/13424553/stef.zip)

**Comment by olisnr** [comment]
thanks for fixing! 

this was stupid! i didnt think about it, i was only trying learn TCL...

what do You think about the memory? do You think there is a problem? if i repeat the same simulation many times, i would expect that the memory-usage stays at some point constant. may be 2 times the sim-data or so.

**Comment by StefanSchippers** [comment]
after the fix the raw files are no more corrupt and I don't see memory issues.
you can run xschem in full debug:

xschem stef.sch -d 3 -l ~/log

load and unload the sim data several tines, then quit the program, then go into the xschem sources src/ directory and run:
./track_memory.awk ~/log
you get:
```
schippes@asus:~/xschem-repo/trunk/src$ ./track_memory.awk ~/log
peak allocated memory = 76508366
Total allocated memory after exit = 0
# of malloc's = 192731
# of realloc's = 367867
# of free's = 257441
Number of unfreed pointers = 0
Total leaked memory = 0
```
so a maximum of 76.5MB was allocated (process size is bigger since it includes all the code and library data), all the data was released at the end and no dangling pointer was left.

https://github.com/StefanSchippers/xschem/assets/69359491/5708e4f0-9b15-4e9c-90fa-528d4fa83927

**Comment by olisnr** [comment]
if im correct, then You write about loosed handles or pointers to allocated memory. i was writing about that with every simulation, Xschem allocate additional memory. and that even if i reload the schematic, the memory remains allocated (the last version is better!). i wonder why.

i did a boring video to demonstrate it:

https://github.com/StefanSchippers/xschem/assets/120017377/009a8b08-f097-400f-8ee3-690f3ceca965

after a some more simulations:

https://github.com/StefanSchippers/xschem/assets/120017377/95663f33-2c6f-49d6-b752-7a7096d9af4d

**Comment by olisnr** [comment]
now i found that the iHPs openPDK is updated, and i wanted to run some of the test-schematics. i had to change the library-paths a little bit, and was able to simulate and plot. but after loading the plot-data Xschem dosent accept any mouse selection in the schematic any more. also tab-switching dont work. but the menue is ok and also the graphdialog.

i never saw this before. i tried with the openPDK xschemrc and with my own (+ and path-append for the symbols in a click-arrow)


https://github.com/StefanSchippers/xschem/assets/120017377/c0209929-1d56-467b-8fad-cc9e2c599166

tell me if should try something.

**Comment by StefanSchippers** [comment]
Yes this usually happen if a dialog has been left open somewhere, may be obscured...

**Comment by StefanSchippers** [comment]
can you please check this?
`xschem get semaphore`
this should normally return 0
if it is not 0 xschem is half locked, meaning you can not change the circuit select, change tabs etc.
You can force this to be zero and xschem should then work as usual.
`xschem set semaphore 0`
Problem is understanding what caused the semaphore to remain "red" (1)

**Comment by olisnr** [comment]
```
name=h1
descr="load sym Ctrl + left click" 
tclcommand="
append XSCHEM_LIBRARY_PATH :/ALL/Xschem/IHP-Open-PDK-main/ihp-sg13g2/libs.tech/xschem
"
```

**Comment by olisnr** [comment]
```
% xschem get semaphore
1

```

yes somebody blocks it after loading the dc-sim

**Comment by StefanSchippers** [comment]
Can you prepare a minimal test case? I have prepared a schematic with the identical issues as yours, a missing symbol, a launcher that fixes the path, descending and returing from a vsource to reload the schematic, double click on the netlist/simulate/wait/load launcher, closing the dialog box. --> everything works.

Can not get it to fail.

**Comment by StefanSchippers** [comment]
> if im correct, then You write about loosed handles or pointers to allocated memory. i was writing about that with every simulation, Xschem allocate additional memory. and that even if i reload the schematic, the memory remains allocated (the last version is better!). i wonder why.

Operating systems are designed to deal with these situations. In many cases if a program frees memory it is not returned to the OS, because it is assumed it will need it again. Subsequent allocations will be faster because the memory pages are already there. No need to search for free memory pages.

Quoting a comment from someone that surely knows things better than me:
_**free often doesn't return memory to the operating system because it assumes that your program will need it again in the future and it's faster to just keep the memory around inside the program. – 
[Art](https://stackoverflow.com/users/939410/art)**_

It is fair to assume that if the OS runs short on memory the freed chunks will be reclaimed from processes.

**Comment by olisnr** [comment]
would it be good to start the xschem file with `#!/usr/local/bin/xschem` ?

**Comment by olisnr** [comment]
i will try tomorrow to make this test-case

**Comment by StefanSchippers** [comment]
> would it be good to start the xschem file with `#!/usr/local/bin/xschem` ?

I never tested this. 
It seems if you put the above hashbang on the first line of a .sch file , make it executable you can call the .sch file directly. 
seems working, but never tested and used...

**Comment by olisnr** [comment]
the question is, if You want to use some mechanism to link the file to Xschem? optimal also in windows (even if i dont care about it). then there are many different `*.sch` files and simply use the extension isnt usable.

i like to double-click in a file-browser to start the schematics.

**Comment by StefanSchippers** [comment]
when a schematic file begins with #!/path/to/xschem
the shell recognizes it is a script, starts the specified program (xschem) and gives the entire file as argument to the program.
Since the shebang starts with # and xschem considers lines beginning with # as comments it is safe to shebang schematic files.
However the shebang is lost when saving.

**Comment by olisnr** [comment]
it could also be an TCL-script that loads it self in Xschem

**Comment by StefanSchippers** [comment]
> the question is, if You want to use some mechanism to link the file to Xschem? optimal also in windows (even if i dont care about it). then there are many different `*.sch` files and simply use the extension isnt usable.
> 
> i like to double-click in a file-browser to start the schematics.

If this is what you want to do then the shebang is not the right way to go. You need to register a new file type with the default application (xschem). May be some info can be found [here](https://stackoverflow.com/questions/30931/register-file-extensions-mime-types-in-linux).

**Comment by olisnr** [comment]
about the memory i think: if the OS is not taking back freed memory, because it thinks that the application will need it again, then Xschem should not allocate new memory from the OS, because it has already more then it needs.
we load always the same size of data, and free also that much. only some handles for ngspice and so on will grow. may be i make an TCL-loop that loads and unloads the hole time an look, if i overflow the RAM at some point.

**Comment by StefanSchippers** [comment]
I have done a test (on a smaller test case) to load/unload raw  files 1200 times, no significant memory increase was observed (just a small quantity, but may be glibc or tcl buffers xschem does not control directly).

https://github.com/StefanSchippers/xschem/assets/69359491/baf11e29-b841-47c2-aba8-1ac0b95815d0

**Comment by olisnr** [comment]
after 1 sim
![mem 1](https://github.com/StefanSchippers/xschem/assets/120017377/68d8248e-b1cd-4e55-88cc-ea5dffc2dc3d)

after 2 sim
![mem 2](https://github.com/StefanSchippers/xschem/assets/120017377/858a6065-5db3-4786-be0b-5dc0472cc83d)

after 32 sim
![mem 32](https://github.com/StefanSchippers/xschem/assets/120017377/6513afe5-2504-4d74-84f8-48bcbefb1a3f)

after 54 sims Xschem has 3GB... i look if it stops somewhere

**Comment by olisnr** [comment]
here comes the testcase: i started **Xschem** from the dirctory with the two `.sch` files.

then i clicked "Netlist", "Simulate" and then "Waves - Dc" then i clicked "Ok". the waveforms are loaded, but the gods of the semaphore arent happy whit it...

[blocktest.zip](https://github.com/StefanSchippers/xschem/files/13438986/blocktest.zip)

**Comment by StefanSchippers** [comment]
> here comes the testcase: i started **Xschem** from the dirctory with the two `.sch` files.
> 
> then i clicked "Netlist", "Simulate" and then "Waves - Dc" then i clicked "Ok". the waveforms are loaded, but the gods of the semaphore arent happy whit it...
> 
> [blocktest.zip](https://github.com/StefanSchippers/xschem/files/13438986/blocktest.zip)

Unfortunately I can not trigger the problem. May be  you have some specific xschemrc setting? (there was no xschemrc so i created one with your defauts).


https://github.com/StefanSchippers/xschem/assets/69359491/cb6e3ec2-1ce3-47cc-8597-308aeee8a809

**Comment by olisnr** [comment]
my settings:

```
set replace_key(Control-z) u        ;# undo
set replace_key(Control-y) Shift-U  ;# redo
set replace_key(z) Control-z        ;# zoom out
set unzoom_nodrift 1
set draw_crosshair 1
set persistent_command 1 ;#for click instead of w

#set enable_stretch 1 ;#for preserve the connections
#set connect_by_kissing 1 ;#for preserve the connections
set unselect_partial_sel_wires 1 ;#for preserve the connections

set local_netlist_dir 1


#append XSCHEM_LIBRARY_PATH :/ALL/Xschem
append XSCHEM_LIBRARY_PATH :/ALL/Xschem/symbol
append XSCHEM_LIBRARY_PATH :/ALL/Xschem/design
append XSCHEM_LIBRARY_PATH :/usr/local/share/xschem/xschem_library


set no_ask_save 1

set editor {xed -w}

set terminal {gnome-terminal }

set cadgrid 80

set autotrim_wires 1
set incremental_select 0
set zoom_full_center 1

set tabbed_interface 0
```

**Comment by olisnr** [comment]
https://github.com/StefanSchippers/xschem/assets/120017377/b6379c7a-f180-4014-9723-aadcfa54dd9a

**Comment by StefanSchippers** [comment]
Thank you for the test case and the update video. Looking to find the culprit. Is your installed xschem the last one? just to avoid research with older commits

**Comment by olisnr** [comment]
its the commit from today (-13h)

**Comment by StefanSchippers** [comment]
Please create an `xschemrc` file  in the `blocktest/` directory, containing this:
```
set replace_key(Control-z) u        ;# undo
set replace_key(Control-y) Shift-U  ;# redo
set replace_key(z) Control-z        ;# zoom out
set unzoom_nodrift 1
set draw_crosshair 1
set persistent_command 1 ;#for click instead of w
#set enable_stretch 1 ;#for preserve the connections
#set connect_by_kissing 1 ;#for preserve the connections
set unselect_partial_sel_wires 1 ;#for preserve the connections
set local_netlist_dir 1

set XSCHEM_LIBRARY_PATH {}
append XSCHEM_LIBRARY_PATH :[pwd]
append XSCHEM_LIBRARY_PATH :$XSCHEM_SHAREDIR/xschem_library

set no_ask_save 1
set editor {xed -w}
set terminal {lxterminal}
set cadgrid 80
set autotrim_wires 1
set incremental_select 0
set zoom_full_center 1
set tabbed_interface 0
```

This way when xschem starts this file will be used and not any other.
Does the 'semaphore' problem still occur?

**Comment by olisnr** [comment]
its still blocked

**Comment by StefanSchippers** [comment]
I think last commit fixes the problem. This was **very hard to spot** since it was a combination of different things (including 4 lines deleted in xschem.tcl by mistake , probably due to the "interrupted while doing a change" effect and the change got done halfway :-) )

There was a flaw in the "configure simulators and tools"

What you need to do:

- If there is a file `~/.xschem/simrc` rename to something different, like `~/.xschem/simrc_old`.
- start xschem and reconfigure `Simulation->Configure simulators and tools`. If you need you can look in the `simirc_old` for command lines. Do not keep the old `simrc` file since it is probably broken.

I think the semaphore locking will no more occur. Please if you have a chance to test let me know.
Thank you for your patience.

**Comment by olisnr** [comment]
Super! its working!

**Comment by olisnr** [comment]
something other is strange: if i open a sub-sheet via `alt-e`, and open a dialog-box in this sub-sheet, then the focus jumps  back to the top-sheet, if i click into the dialogbox.

https://github.com/StefanSchippers/xschem/assets/120017377/3a0ef5cf-2dce-49c4-9168-9fe28dbd6041

and also the mouse is a bit strange:

https://github.com/StefanSchippers/xschem/assets/120017377/1cb1425d-99ee-4f64-a98f-84f8d5ef14f3

**Comment by StefanSchippers** [comment]
Thank you for reporting this. In multi window interface now clicking on an element to edit its attributes (as well as many other dialog boxes) will no more raise the top sheet. Please test.

For the second problem, how windows are focused (and how windows loose focus) is decided by the window manager. Can you please click into a window and see if the mouse behaves correctly? 
some focus policies require a click to activate a window (Click to focus) some others not (Focus follows Mouse) Modern window managers have then a plethora of other fancy options, may be the next generation WMs  will use AI to decide.

**Comment by StefanSchippers** [comment]
In general multi window interface is not recommended. It is MUCH better to fire another xschem instance if you want another separate window (or use tabbed interface).
A single xschem process handling multiple windows is tricky as it must switch context to the right window whenever the mouse flies over a window  or enters / exits a window. Since the application <--> Xserver is asynchronous there are many corner cases that can lead to malfunction. It is complex since it is a threesome: Application - Xserver - Window manager.
In tabbed interface the switchover is simpler and explicit. You click the tab.

**Comment by olisnr** [comment]
_... It is MUCH better to fire another xschem instance ..._ what does this mean? open the sub-sheet in a new Xschem process? dont i do this using the anti-tab interface?

**Comment by olisnr** [comment]
the "jiump to top" problem is fixed. but i also think sometimes the Tk isnt sooo perfect matched to the fancy new window-managers:
evey time the mouse is not over the text-field from a dialog-box, the text-field looses the keybord-inputs. other programs like `xed` dont do this. You can go with the mouse on even some other screens, and as long You dont click somewhere, all keyboard input is included.

**Comment by StefanSchippers** [comment]
> _... It is MUCH better to fire another xschem instance ..._ what does this mean? open the sub-sheet in a new Xschem process? dont i do this using the anti-tab interface?

exactly. In windowed interface you are not creating a new process. Xschem just creates a new toplevel window.
It is possible to open a subsheet in a different process with Alt-Shift-E

**Comment by StefanSchippers** [comment]
> the "jiump to top" problem is fixed. but i also think sometimes the Tk isnt sooo perfect matched to the fancy new window-managers: evey time the mouse is not over the text-field from a dialog-box, the text-field looses the keybord-inputs. other programs like `xed` dont do this. You can go with the mouse on even some other screens, and as long You dont click somewhere, all keyboard input is included.

The focus policy is set by the window manager, either "focus follows mouse" or "click to focus".
in "click to ficus" mode, both on IceWM and FVWM the tk text boxes behave exactly like other applications.
Some window managers can apply different policies to different applications, but this depends on the WM.

https://github.com/StefanSchippers/xschem/assets/69359491/e1ac7320-2107-4638-934f-8c555e4fe7f4

**Comment by StefanSchippers** [comment]
Oh yes, there is an option (was asked by someone don't remember):

`set autofocus_mainwindow 0`

try to add this line.

from the commit history:
"_added xschemrc variable `autofocus_mainwindow` (default setting: enabled (1) ), if this variable is set moving the mouse to the main drawing area when a dialog box is open (typically while editing a component attribute list) will focus the main window, so user can move / zoom/pan the schematic using the usual commands, for example to look at a different part of the schematic while editing the component. If this variable is unset a click is needed to focus the main window. This will thus not allow to move / zoom, but allows to type into the dialog box if the mouse goes out of it_."

**Comment by olisnr** [comment]
great. it works!

**Comment by StefanSchippers** [comment]
Sorry i did nott remember this setting. By the way I also use this `set autofocus_mainwindow 0` because the textboxes losing focus if you push the mouse with the arm while typing is annoying.

**Comment by olisnr** [comment]
yes, but navigating across the schematic while a text-dialog is open is also great. we need the option: 
`set autofocus_mainwindow_while_zooming_via_mouse_wheel_around 0`! and my experience with Xschem, says me this option is allready implementet! (i will try it out)

**Comment by olisnr** [comment]
by the way, what settings are You using Yourself?

**Comment by olisnr** [comment]
if i have more the 2 Xschem as there own process the copy-function does work a bit strange: the part is captured in the first window i move over.


https://github.com/StefanSchippers/xschem/assets/120017377/e904e248-6aa3-4a6e-8deb-924af2a3e364

**Comment by StefanSchippers** [comment]
Yes this is done on purpose, to allow copying pieces across windows. If this is not desired move the other window apart or minimize it.

**Comment by olisnr** [comment]
copy across windows is very desired and it would be interesting to be able to copy to any window: 
if i have 4 windows open (in a gid of 2 x 2), its only possible to copy to 2 of them. im not able to copy to the window thats diagonal from the source. its also strange, thats not possible to take the part back in to the source-window, but only if i was touching a 2nd target window.
my work around is `cntr-c` `cntr-v`

**Comment by StefanSchippers** [comment]
Yes the communication between the different windows is limited. I will look if there is a way to improve that. I also found some crashes while aborting (with escape) such operations, so I have something to do in the afternoon :-)

**Comment by StefanSchippers** [comment]
The occasional crash problem (when copying between windows) is fixed. Will look if there is a simple way for each target window to grab the copy (including the window the copy started from).

**Comment by olisnr** [comment]
now i had a crash while autorangeing (f - key) the y axis of a plot. Xschem said:

```
...
Raw file data read: /ALL/Xschem/design/simulation/aa_pmosfets_m_factor.raw
points=1, vars=282, datasets=1 sim_type=op
Raw file data read: /ALL/Xschem/design/simulation/aa_pmosfets_m_factor_dc.raw
points=38507, vars=283, datasets=1 sim_type=dc
Raw file data read: /ALL/Xschem/design/simulation/aa_pmosfets_m_factor_ac.raw
points=61, vars=566, datasets=1 sim_type=ac
free_rawfile(): clearing data
free_rawfile(): clearing data
free_rawfile(): clearing data
Raw file data read: /ALL/Xschem/design/simulation/aa_pmosfets_m_factor.raw
points=1, vars=282, datasets=1 sim_type=op
Raw file data read: /ALL/Xschem/design/simulation/aa_pmosfets_m_factor_dc.raw
points=38507, vars=283, datasets=1 sim_type=dc
Raw file data read: /ALL/Xschem/design/simulation/aa_pmosfets_m_factor_ac.raw
points=61, vars=566, datasets=1 sim_type=ac
done
EMERGENCY SAVE DIR: /tmp/xschem_emergencysave_aa_pmosfets_m_factor_eeafadebag

FATAL: signal 11
while editing: aa_pmosfets_m_factor
```

i have plot crashes from time to time, but i cant reproduce it. sometimes it crashes also in the plotdialog.

**Comment by StefanSchippers** [comment]
The copy between windows is now improved. please test.
Thank you for the info about the crash while zooming. If I can trigger the crash it would be easy to fix.

**Comment by olisnr** [comment]
what is in the `xschem_emergencysave_aa_pmosfets_m_factor_eeafadebag`? is it helpful?

**Comment by olisnr** [comment]
now i had the same with auto-ranging the x-axis. load the .raw files and autorange:

[aa_pmosfets_m_factor.sch.zip](https://github.com/StefanSchippers/xschem/files/13491422/aa_pmosfets_m_factor.sch.zip)

**Comment by olisnr** [comment]
copy via `c` works great. 

there remain are some "ghost" devices on the window, that looses there focus, maybe this could also be removed in the correct event. 

but it works!

**Comment by StefanSchippers** [comment]
> what is in the `xschem_emergencysave_aa_pmosfets_m_factor_eeafadebag`? is it helpful?

In case of a crash a signal handler is called. If the signal handler is able to complete its duty, the undo stack is saved to /tmp/xschem_emergencysave_xxxxxx directory.

The directory contains all undo levels (up to MAX_UNDO, 80) present when the program crashed.
these undo1, undo2, .... undon files inside the directory are just gzip compressed xschem files.

you can restore one such undo by doing:

`zcat /tmp/xschem_emergencysave_aa_pmosfets_m_factor_eeafadebag/undox > ~/xxx.sch`

you can then open xxx.sch and if it looks ok you get your schematic back. 
You should do a `ls -ltr /tmp/xschem_emergencysave_aa_pmosfets_m_factor_eeafadebag` to see the last undo that was written. This is the most recent schematic you were working on when xschem crashed.

**Comment by StefanSchippers** [comment]
> now i had the same with auto-ranging the x-axis. load the .raw files and autorange:
> 
> [aa_pmosfets_m_factor.sch.zip](https://github.com/StefanSchippers/xschem/files/13491422/aa_pmosfets_m_factor.sch.zip)

Thank you I hope to be able to trigger the failure so I can fix that. If you start the schematic and just load the raw files (no simulations) do you see the crashes? or does this happen only if you run the simulations before loading the graphs?

**Comment by olisnr** [comment]
no: i open the schematic, then i load the .raw-data via the "Wave - OP annotate", and the i try to autorange different axis from different plots until it crashes.
may be You have to reload if it dont crashes fast enough. on my laptop it crashes so all 6 tries or so...

**Comment by StefanSchippers** [comment]
> now i had the same with auto-ranging the x-axis. load the .raw files and autorange:
> 
> [aa_pmosfets_m_factor.sch.zip](https://github.com/StefanSchippers/xschem/files/13491422/aa_pmosfets_m_factor.sch.zip)

I am looking at the simulation commands in the schematic (I can not run the sims since I have not the pdk/model data)

This loop:
```
	foreach corner $corners {
		set hv_corner $corner; 
		xschem netlist; 
		delay 111;
		set [string cat {id_} ${corner}] [simulate]; 
	}
```

should be modified by puttint the delay 111 after the [simulate] line
The reason is you must give some time to ngspice to read and parse the netlist before generating a new netlist.
I also recommend using 300ms at least to ensure ngspice has finished reading the netlist:

```
	foreach corner $corners {
		set hv_corner $corner; 
		xschem netlist; 
		set [string cat {id_} ${corner}] [simulate]; 
                delay 300
	}
```

PS: In TCL you don't need to close each line with **;**. **;** is used to separate commands on the same line, like in bash

**Comment by StefanSchippers** [comment]
Good news. I also had a crash while full-zooming graphs. I think I can start debugging.

**Comment by StefanSchippers** [comment]
I think it is fixed now. 
graph_fullxzoom() function did look up the sweep variable (the x axis) position in raw file before switching to the graph specific raw file, so looking in a raw file that is potentially not the one used in the graph.
Can you please try to crash again with the autorange plots?
Let me know if you see differences. I do no more see crashes now.
Thank you for your cooperation.

**Comment by olisnr** [comment]
> should be modified by putting the delay 111 after the [simulate] line The reason is you must give some time to ngspice to read and parse the netlist before generating a new netlist. I also recommend using 300ms at least to ensure ngspice has finished reading the netlist:
> 
> ```
> 	foreach corner $corners {
> 		set hv_corner $corner; 
> 		xschem netlist; 
> 		set [string cat {id_} ${corner}] [simulate]; 
>                 delay 300
> 	}
> ```

You are right, the delay is at the wrong position. but it would be saver to write different netlists. i will try this.

**Comment by olisnr** [comment]
no i had a "undo" button from an Xschem instance on the screen. i made a video, but the button was from the other instance. it disappeared with closing the 2nd Xschem. i hope i can reproduce it...

https://github.com/StefanSchippers/xschem/assets/120017377/02a90787-3027-4a44-b25a-f923c35b934d

**Comment by olisnr** [comment]
is it possible to change the `netlist_name` in a `launcher.sym`? i want to simulate different netlists for different simulations and need a way to tell ngspice the netlist_name.  but

```
	global netlist_name

	foreach corner $corners { # simulate all netlists
		set netlist_name [rawfilestart]_mos${corner}.spice
		set id_${corner} [simulate ]
		puts $netlist_name # debug
	}
```

didnt work. ngspice always uses the standard netlist-name.

**Comment by StefanSchippers** [comment]
This is the Undo tooltip. If the tooltip remains open you must hover the mouse over the undo toolbar button of the xschem instance that generated it. It has happened to em also once. Difficult to reproduce.

**Comment by olisnr** [comment]
something i really dont get is the handling of `"` in xschems evaltcl. if i try some lines in tclsh:
```
    puts test
    puts {test test}
    puts "test test test"
    puts [return "test test test test"]
    puts [subst "test test test test test"]
    puts "\"test test test test test test test\""
    puts eval [return "test test test test test test test test"]
```
i get
```
test
test test
test test test
test test test test
test test test test test
"test test test test test test test"
test test test test test test test test
```
as result.

but in Xschem i get for:

```
value="tcleval(
    test
    {test test}
    "test test test"
    [return "test test test test"]
    [subst "test test test test test"]
    "\"test test test test test test test\""
    eval [return "test test test test test test test test"]
)"
```
this:
![tcl](https://github.com/StefanSchippers/xschem/assets/120017377/d657fa89-e82d-461f-af4f-28ad483c4178)

is this correct? and how is it possible to print a substituted line (or even multiline) in `tcleval() ` ?

**Comment by StefanSchippers** [comment]
To test what you get from a tcleval() you can use the tcl subst command:

the displayed values are correct:

the sequence:
```
value="tcleval(
    test
    {test test}
    "test test test"
    [return "test test test test"]
    [subst "test test test test test"]
    "\"test test test test test test test\""
    eval [return "test test test test test test test test"]
)"
```

is interpreted as:
value="tcleval(
    test
    {test test}
    "test

the double quote closes the opening double quote. Since the following test word has no spaces it is also considered. Anything following is discarded. If you want to use double quotes inside a "tcleval(.....): tou must escape them:

If you use this:
```
value="tcleval(
    test
    {test test}
    \"test test test\"
    [return \"test test test test\"]
    [subst \"test test test test test\"]
    \"test test test test test test test\"
    eval [return \"test test test test test test test test\"]
)"
```
you get
![1](https://github.com/StefanSchippers/xschem/assets/69359491/b5be0328-1cfd-486e-941e-4fbce8a6657b)

**Comment by StefanSchippers** [comment]
Remember subst (the command that is applied on the string inside tcleval(...) does only variable ($xxx) and command ([cmd]) substitution, anything else is returned as is. 
So a tcl command like `eval` or `puts` is not considered special by subst, it becomes a command to be substituted if it appears inside [...]

**Comment by StefanSchippers** [comment]
If you find other usages where the stupid thing (the tk_messageBox) causes problems let me know, I am slowly getting rid of it everywhere. Unfortunately there are many such messageboxes around.

**Comment by olisnr** [comment]
sometimes the "File" menue is on the other screen, but only on the left one.

![menue](https://github.com/StefanSchippers/xschem/assets/120017377/5c9615b8-794c-4357-9eb2-65736c61f9c2)

**Comment by olisnr** [comment]
i played a bit around, but i have no idea how to force this. normally the menus open under the mouse click...

**Comment by StefanSchippers** [comment]
May be the menu goes in the other screen if the left border of the xschem window is outside the monitor and goes (may be by one pixel) in the other screen? Don't know.

**Comment by StefanSchippers** [comment]
I remember you asked a way to generate netlists with different names, I was looking at it but can't find the post. Can you explain again?

**Comment by StefanSchippers** [comment]
Don't know if you find this interesting.
the `set local_netlist_dir 1` will create a `simulation/` directory where netlists and simulation data will go.
There is now another setting: `set local_netlist_dir 2`. This will add one directory level as follows:
`library_dir/simulation/schematic_name/`

example:
```
schippes@asus:~/.xschem/xschem_library/olisnr2$ ls -l simulation
total 12
drwxr-xr-x 2 schippes schippes 4096 Dec  1 17:38 aa_opamp_p_test03.sch
drwxr-xr-x 2 schippes schippes 4096 Dec  1 17:38 aa_opamp_p_testi03.sch
drwxr-xr-x 2 schippes schippes 4096 Dec  1 17:38 ihp_mos_gms_ngspice_05.sch
```
This way all simulation data will be in separate directories. Beware if you do a `.include ../something` this will probably need to be changed to `.include ../../something`.

**Comment by olisnr** [comment]
yes thats interesting. it should go into xschemrc?

**Comment by StefanSchippers** [comment]
Yes,
set local_netlist_dir 2
Or in Simulation menu.

**Comment by StefanSchippers** [comment]
for lines and wires when in `persistent command mode` the first escape clears the ongoing placement, the second escape ends the lining or wiring command.
The delete function requires a little more work.

**Comment by olisnr** [comment]
> > instead of `esc` You also could use `right-click` (would be even better) but then there could be a conflict with the context-menue.
> 
> Left button Double-Click does exactly that.

yes. i didnt got it. its exactly what i wanted.  

the command as procedure is also working now.

**Comment by olisnr** [comment]
can i also get rid of the messageboxes about the missing `symbols.sch` from `symbols.sym`, that have a `schematic` attribute?

![missing sch](https://github.com/StefanSchippers/xschem/assets/120017377/8ce54d97-9002-4ee0-954d-12d61132ab83)

**Comment by olisnr** [comment]
there is an other thing with the spice netlist. for some reason the line-lenght is limited. and this makes problems with the write command. as example the first is working but the second not:

```
                writes  /ALL/Xschem/design/simulation/aa_opamp_dual_supply_RjBaker_05_zoom_dc_mostt.raw
                ...
```
```
                writes 
+ /ALL/Xschem/design/simulation/aa_opamp_dual_supply_RjBaker_05_zoom_dc_mostt.raw
                ...
```

**Comment by StefanSchippers** [comment]
Thank you this is now fixed. User code is not touched now. Xschem needs to run a line breaker at the very end of the netlist generation, because many simulators have a maximum line length limit (and digital blocks with [1023:0] buses can generate very long lines). But breaking code written by the user is not good.

**Comment by olisnr** [comment]
> ... But breaking code written by the user definitely not good. I have also incremented the line length from 100 to 130 (less breaks)

yes i think its better to let the user limit the line-size, because i had also problems using a spice library with many + linebreaks: Xschem made a mess with it, so that i had to remove all "\n+" and replace it with " ", to allow Xschem to brake the code it self. this isnt necessary further.
and why did You change to 130?

**Comment by StefanSchippers** [comment]
the format is :
`"alias_name; expression [% [dataset] rawfile sim_type]"`
Example:
`"del; diffout 0.1u del() % 0 /home/schippes/.xschem/simulations/cmos_example_ngspice.raw tran"`

It is not possible to pick one node from one file and do math operations with another node from another file, since the number of points is different (and on top of that transient timesteps are not uniform, it is very difficult to do).

Calculations on graphs are done point by point

![1](https://github.com/StefanSchippers/xschem/assets/69359491/b0ee0502-ae7a-4891-956e-9d65595e6f56)

**Comment by StefanSchippers** [comment]
> and why did You change to 130?

previous limit was 100 and it seems too low. Using 130 will less likely introduce breaks for reasonable wide texts.
Anyway the updated xschem will not touch any more texts brought into the netlist by code.sym elements.

**Comment by olisnr** [comment]
> Of course it should be possible to analyze the full hierarchy and at the end decide what to do. This requires a 2 pass full hierarchy traverse and this is Cadence-ish style, very inefficient.

thats a good reason...

**Comment by olisnr** [comment]
the `!` command dont work correct on multi-selections. only the wires that crosses the pins of one part are cut. 
and would it be ok to remove the dialogbox after the `y` command? if the user should be aware of the mode, maybe it would be better to put the information in the status line at the bottom of the window?

**Comment by StefanSchippers** [comment]
cutting on multiple selections just works.

https://github.com/StefanSchippers/xschem/assets/69359491/65b9b4e9-1210-4629-b838-0268af00c6ce

**Comment by olisnr** [comment]
yes it is strange. i saw this a few times, when i wanted to cut "antennas" after moving transistors around, but now i cant reproduce it reliable. it did show up only one time, while i tried to find a case i could record...
may be it appears in future, and then i will report. 
maybe i have to verify that all things are on a the grid?

**Comment by StefanSchippers** [comment]
> would it be ok to remove the dialogbox after the `y` command?

This is now done. info shown in status bar. Thank you.

**Comment by olisnr** [comment]
> One of the reason is if devices are (even by s small amount) off grid

it happens again a few times, but only after some time of editing, and i couldent repeat it. most of the time i used `cntr-a` first, and then `!`. one time i looked in the text of the `*.sch` file, but found only parts and wires in decadic-coordinates. the thing is: if i save the schematics and reload, it works again...

**Comment by StefanSchippers** [comment]
> > One of the reason is if devices are (even by s small amount) off grid
> 
> it happens again a few times, but only after some time of editing, and i couldent repeat it. most of the time i used `cntr-a` first, and then `!`. one time i looked in the text of the `*.sch` file, but found only parts and wires in decadic-coordinates. the thing is: if i save the schematics and reload, it works again...

I found the reason you sometimes don't see the wire removal.
If wires are already broken at the symbol pins the inner part is not removed. Will see if I can fix that.
In picture below the wire under the lower inductor will not be removed.
![1](https://github.com/StefanSchippers/xschem/assets/69359491/ac0b94d6-c9e4-453a-9810-54538456adb7)

**Comment by olisnr** [comment]
my cases are a bit different, i typical see it on the gate of MOS transistors, if i want remove the "anntenas":

![mycase](https://github.com/StefanSchippers/xschem/assets/120017377/918f264d-cf79-485a-9285-2e3ed5336efd)

**Comment by StefanSchippers** [comment]
I have improved a bit the 'antenna cutter' so wires touching pins all inside symbols will be deleted, as well as wires all inside touching one pin and touching nothing on the other side.
![1](https://github.com/StefanSchippers/xschem/assets/69359491/53386051-fdd2-485e-969a-231f1f703d97)

**Comment by olisnr** [comment]
is there a change in the trafficlight of newer versions? at a click to "Netlist" the "Simulation" looses the color, but the "Wave" not. if green means that waves are loaded, then maybe it would be more clear to switch "Wave" to yellow, showing, that the waves that are loaded are from an old simulation?

**Comment by StefanSchippers** [comment]
I have made the change.
If waves are loaded (Green) then Waves button will turn to yellow upon pressing 'Netlist'.

**Comment by olisnr** [comment]
did You think about adding a vertical flip? (1 time flip and 2 times rotate). and would it be ok to set the cursor at opening the "Choose symbol" dialog to "File/Search"?

**Comment by StefanSchippers** [comment]
Vertical flip is done. `Shift-V`  
Vertical flip in place is done. `Alt-v`
The "change netlisting mode" command has been moved to key `ctrl-shift-V` 
Also when Insert component dialog is invoked focus the FIle/Search entry box.

Please test! :-)

**Comment by olisnr** [comment]
tanks! works great! & Happy New Year!
 
i think it would be a good idea to put the `property * toggle` into the context menue for devices. and if the plots dont have an file specified i would expect the same `.raw` name like in the `-r "$n.raw"` from the simulation command dialog.

**Comment by olisnr** [comment]
now i have a window with a "gost schematics"... i copy and pasted some parts... i never saw this before...

https://github.com/StefanSchippers/xschem/assets/120017377/ab8587fc-2952-4979-879c-ef3c25d99ec4

**Comment by olisnr** [comment]
yes! thats it! thanks

**Comment by StefanSchippers** [comment]
It's a kind of 'graphical diff' just in case you have 2 versions of the same schematic and you want see where the differences are.
